homeassistant/components/weatherkit/config_flow.py

```diff
@@ -0,0 +1,122 @@
+"""Adds config flow for WeatherKit."""
+from __future__ import annotations
+
+from collections.abc import Mapping
+from typing import Any
+
+from apple_weatherkit.client import (
+    WeatherKitApiClient,
+    WeatherKitApiClientAuthenticationError,
+    WeatherKitApiClientCommunicationError,
+    WeatherKitApiClientError,
+)
+import voluptuous as vol
+
+from homeassistant import config_entries, data_entry_flow
+from homeassistant.const import CONF_LATITUDE, CONF_LOCATION, CONF_LONGITUDE, CONF_NAME
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.selector import (
+    LocationSelector,
+    LocationSelectorConfig,
+    TextSelector,
+    TextSelectorConfig,
+)
+
+from .const import (
+    CONF_KEY_ID,
+    CONF_KEY_PEM,
+    CONF_SERVICE_ID,
+    CONF_TEAM_ID,
+    DOMAIN,
+    LOGGER,
+)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME, default="Home"): str,
+        vol.Required(CONF_LOCATION): LocationSelector(
+            LocationSelectorConfig(radius=False, icon="")
+        ),
+        # Auth
+        vol.Required(CONF_KEY_ID): str,
+        vol.Required(CONF_SERVICE_ID): str,
+        vol.Required(CONF_TEAM_ID): str,
+        vol.Required(CONF_KEY_PEM): TextSelector(
+            TextSelectorConfig(
+                multiline=True,
+            )
+        ),
+    }
+)
+
+
+class WeatherKitUnsupportedLocationError(Exception):
+    """Error to indicate a location is unsupported."""
+
+
+class WeatherKitFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Config flow for WeatherKit."""
+
+    VERSION = 1
+
+    async def async_step_user(
+        self,
+        user_input: dict | None = None,
+    ) -> data_entry_flow.FlowResult:
+        """Handle a flow initialized by the user."""
+        _errors = {}
+        if user_input is not None:
+            try:
+                await self._test_config(user_input)
+            except WeatherKitUnsupportedLocationError as exception:
+                LOGGER.error(exception)
+                _errors["base"] = "unsupported_location"
+            except WeatherKitApiClientAuthenticationError as exception:
+                LOGGER.warning(exception)
+                _errors["base"] = "auth"
+            except WeatherKitApiClientCommunicationError as exception:
+                LOGGER.error(exception)
+                _errors["base"] = "connection"
+            except WeatherKitApiClientError as exception:
+                LOGGER.exception(exception)
+                _errors["base"] = "unknown"
+            else:
+                return self.async_create_entry(
+                    title=user_input[CONF_NAME],
+                    data=user_input,
```

Comment:
We like our config entry data structures flat, so it should flat out the CONF_LOCATION so the latitude and longitude are on the same level as the rest

Reply:
Got it. I don't think this was mentioned in the developer docs. Is there any particular reason for this? Easier to serialize?

Reply:
Well, it would certainly work, I also got this as feedback as one of my own PRs, but I am not sure about the reasoning.

I think the main goal would be to avoid integration putting in complete data structures.

In this case, accessing one level down costs more than staying on the same level. On this scale it doesn't matter that much, but when you get complete data structures in a lot of different integrations... it would get out of hand.

Note: we do use an object in oauth integrations to store token related data. Why this is chosen, not sure, but just to let you know that the story above isn't 100% applicable.